syntax = "proto3";

package gateway;

// Mensagens para o Gateway (agregação de todos os serviços)
import "leilao.proto";
import "lance.proto";
import "pagamento.proto";

// Notificações unificadas
message NotificacaoUnificada {
    string tipo = 1; // "leilao_iniciado", "novo_lance_valido", "vencedor_leilao", "link_pagamento", etc
    int32 leilao_id = 2;
    string user_id = 3;
    double valor = 4;
    string id_vencedor = 5;
    string link_pagamento = 6;
    string status = 7;
    string message = 8;
    leilao.Leilao leilao = 9;
}

message StreamNotificacoesUnificadasRequest {
    string cliente_id = 1;
}

// Serviço Gateway que agrega todos os outros
service GatewayService {
    // Leilão
    rpc CriarLeilao(leilao.CriarLeilaoRequest) returns (leilao.CriarLeilaoResponse);
    rpc ListarLeiloes(leilao.ListarLeiloesRequest) returns (leilao.ListarLeiloesResponse);
    rpc RegistrarInteresse(leilao.RegistrarInteresseRequest) returns (leilao.RegistrarInteresseResponse);
    rpc CancelarInteresse(leilao.CancelarInteresseRequest) returns (leilao.CancelarInteresseResponse);
    
    // Lance
    rpc EnviarLance(lance.EnviarLanceRequest) returns (lance.EnviarLanceResponse);
    
    // Pagamento
    rpc ProcessarPagamento(pagamento.ProcessarPagamentoRequest) returns (pagamento.ProcessarPagamentoResponse);
    
    // Stream unificado de notificações
    rpc StreamNotificacoesUnificadas(StreamNotificacoesUnificadasRequest) returns (stream NotificacaoUnificada);
}
